File: 230. Kth Smallest Element in a BST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        self.res = 0
        self.k = k
        self.inorder(root)
        return self.res
    
    def inorder(self, root):
        if not root:
            return
        self.inorder(root.left)
        self.k -= 1
        if self.k == 0:
            self.res = root.val
            return
        self.inorder(root.right)

File: 230. Kth Smallest Element in a BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int res = 0;
        inorder(root, k, res);
        return res;
    }
    
    void inorder(TreeNode* root, int &k, int &res)
    {
        if (root == NULL)
            return;
        inorder(root->left, k, res);
        k--;
        if (k == 0)
        {
            res = root->val;
            return;
        }
        inorder(root->right, k, res);
    }
};
